nixos/modules/services/security/certmgr.nix

```diff
@@ -0,0 +1,200 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.certmgr;
+
+  specs = mapAttrsToList
+    (n: v:
+      let
+        name = n + ".json";
+        path = if isAttrs v then pkgs.writeText name (builtins.toJSON v) else v;
+      in
+        { inherit name; inherit path; })
+    cfg.specs;
+
+  allSpecs = pkgs.buildEnv {
+     name = "certmgr.d";
+     paths = optional (specs != []) (pkgs.linkFarm "certmgr-specs" specs);
+  };
+
+  certmgrYaml = pkgs.writeText "certmgr.yaml" (builtins.toJSON {
+    dir = allSpecs;
+    default_remote = cfg.defaultRemote;
+    svcmgr = cfg.svcManager;
+    before = cfg.validMin;
+    interval = cfg.renewInterval;
+    inherit (cfg) metricsPort metricsAddress;
+  });
+
+  uniqueDirs = list: unique (map (f: escapeShellArg (dirOf f)) (flatten list));
+
+  catDirsRecursive = set:
+  uniqueDirs
+    (mapAttrsToList
+      (n: v:
+          if isAttrs v then
+            collect isString (filterAttrsRecursive (n: v: isAttrs v || n == "path") v)
+          else
+            v)
```

Comment:
Couple things: `n` isn't used -> `concatMap/map` instead. More descriptive name instead of `v` (can be confusing with later `v`). Named this variable properly, and removed the flexibility of `set` and using `cfg.specs` always instead, that's all we need. I don't think using `unique` is worth it, it's an expensive operation (O(n^2)) and mkdir handles duplicates just fine. And some more things. All in all I think it should look like this:

```nix
specPaths = map dirOf (concatMap (spec:
  if isAttrs spec then
    collect isString (filterAttrsRecursive (n: v: isAttrs v || n == "path") spec)
  else
    [ spec ]
) (attrValues cfg.specs)));

preStart = ''
  ${concatStringsSep " \\\n" (["mkdir -p"] ++ map escapeShellArg specPaths)}
  ...
''
```

Reply:
Changed to match your suggestion in: https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224